File: 0457-circular-array-loop/0457-circular-array-loop.py
class Solution:
    def nextIndex(self,i,nums):
        n = len(nums)
        return (i+nums[i])%n

    def circularArrayLoop(self, nums: List[int]) -> bool:
        n = len(nums)

        for i in range(n):

            if nums[i]==0:
                continue
            
            positive = nums[i]>0

            fast =i
            slow = i
            while(True):

                nextSlow = self.nextIndex(slow,nums)

                if nums[nextSlow] == 0 or (nums[nextSlow]>0) != positive or slow == nextSlow:
                    break
                
                slow = nextSlow

                nextFast = self.nextIndex(fast,nums)

                if nums[nextFast] ==0 or (nums[nextFast]>0) != positive or fast == nextFast:
                    break
                
                fast = nextFast

                nextFast = self.nextIndex(fast,nums)

                if nums[nextFast] ==0 or (nums[nextFast]>0) != positive or fast == nextFast:
                    break
                
                fast = nextFast

                if slow == fast:
                    return True
            
            curr = i

            while nums[curr] !=0 and (nums[curr]>0)== positive:
                nxt = self.nextIndex(curr,nums)

                if nums[nxt] == 0 or (nums[nxt]>0) != positive or curr == nxt:
                    break
                nums[curr]=0
                curr = nxt
            
            if nums[curr] >=0 and nums[curr]>0 == positive:
                nums[curr]=0
        return False

File: 0457-circular-array-loop/0457-circular-array-loop.cpp
class Solution {
    int nextIndex(int i , vector<int>& nums){
        int n = nums.size();
        return (( i + nums[i]) % n + n) % n;
    }
public:
    bool circularArrayLoop(vector<int>& nums) {
        int n = nums.size();

        for(int i =0;i<n;i++){

            // already visited 
            if (nums[i]==0){
                continue;
            }

            bool positive = nums[i]>0;

            int fast = i;
            int slow = i;

            while(true){

                int nextSlow = nextIndex(slow,nums);

                // direction must remain same 
                if(nums[nextSlow] == 0 || (nums[nextSlow] > 0 ) != positive){
                    break;
                }

                // self loop
                if(slow == nextSlow){
                    break;
                }

                slow = nextSlow;

                int nextFast = nextIndex(fast,nums);
                if(nums[nextFast] == 0 || (nums[nextFast] > 0) != positive || nextFast == fast){
                    break;
                }
                fast = nextFast;

                nextFast = nextIndex(fast,nums);
                if(nums[nextFast] == 0 || (nums[nextFast] > 0 ) != positive || nextFast == fast){
                    break;
                }

                fast = nextFast;

                if(slow == fast){
                    return true;
                }

                
            }

            // mark path as visited
            int curr = i;

            while(nums[curr]!=0 && (nums[curr]>0) == positive){
                int next = nextIndex(curr,nums);

                if(nums[next] ==0 || (nums[next]>0) != positive || next == curr){
                    break;
                } 

                nums[curr]=0;

                curr = next;
            }

            if(nums[curr] !=0 && (nums[curr]>0) == positive){
                nums[curr] =0;
            }

        }
        return false;
    }
};